src/System.IO.Pipes/src/System/IO/Pipes/PipeStreamCompletionSource.cs

```diff
@@ -0,0 +1,225 @@
+                }
+            }
+        }
+
+        internal void ReleaseResources()
+        {
+            _cancellationRegistration.Dispose();
+
+            // NOTE: The cancellation must *NOT* be running at this point, or it may observe freed memory
+            // (this is why we disposed the registration above)
+            if (Overlapped != null)
+            {
+                _threadPoolBinding.FreeNativeOverlapped(Overlapped);
+                _overlapped = null;
+            }
+        }
+
+        internal void SetCompletedSynchronously(int numBytes = default(int))
+        {
+            if (!_isWrite)
+            {
+                _pipeStream.UpdateMessageCompletion(_isMessageComplete);
+            }
+
+            TrySetResult(numBytes);
+        }
+
+        private void AsyncCallback(uint errorCode, uint numBytes)
+        {
+            _numBytes = (int)numBytes;
+
+            // Allow async read to finish
+            if (!_isWrite)
+            {
+                switch (errorCode)
+                {
+                    case Interop.mincore.Errors.ERROR_BROKEN_PIPE:
+                    case Interop.mincore.Errors.ERROR_PIPE_NOT_CONNECTED:
+                    case Interop.mincore.Errors.ERROR_NO_DATA:
+                        errorCode = 0;
+                        numBytes = 0;
+                        break;
+                }
+            }
+
+            // For message type buffer.
+            if (errorCode == Interop.mincore.Errors.ERROR_MORE_DATA)
+            {
+                errorCode = 0;
+                _isMessageComplete = false;
+            }
+            else
+            {
+                _isMessageComplete = true;
+            }
+
+            _errorCode = (int)errorCode;
+
+            int resultState = errorCode == 0 ? ResultSuccess : ResultError;
+
+            // Store the result so that other threads can observe it
+            // and if no other thread is registering cancellation, continue.
+            // Otherwise CompleteCallback(resultState) will be invoked by RegisterForCancellation().
+            if (Interlocked.Exchange(ref _state, resultState) == NoResult)
+            {
+                // Now try to prevent invocation of CompleteCallback(resultState) from RegisterForCancellation().
+                // Otherwise, thread responsible for registering cancellation stole the result and it will invoke CompleteCallback(resultState).
+                if (Interlocked.Exchange(ref _state, CompletedCallback) != NoResult)
+                {
+                    CompleteCallback(resultState);
+                }
+            }
+        }
+
+        private void Cancel()
+        {
+            // Storing to locals to avoid data races
+            SafeHandle handle = _threadPoolBinding.Handle;
+            NativeOverlapped* overlapped = Overlapped;
+
+            Debug.Assert(overlapped != null && !Task.IsCompleted, "IO should not have completed yet");
```

Comment:
I feel like I may have asked this question before on the other async changes that are similar, but why isn't there a race condition where the Task could be completed here?